bpo-26439 Fix ctypes.util.find_library failure on AIX

aixtools opened the pull request:
This patch implements the function find_library() for AIX.
By default it searches in AIX style archives first, and if None is found, repeats the search looking for a .so file instead. For example:
find_library("c") returns libc.a(shr.o)
while
find_library("rpm") returns librpm.so

If python is a 64-bit build, the routine knows to find the 64-bit members, e.g., 
find_library("c") returns libc.a(shr_64.o)


<!-- issue-number: bpo-26439 -->
https://bugs.python.org/issue26439
<!-- /issue-number -->

aixtools committed:
fix test so that AIX does not traceback by default

aixtools committed:
patch to implement find_library() support in ctypes/util for AIX

aixtools committed:
after patchcheck

aixtools committed:
add a standard .so library to the AIX test in util.py

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

Mariatta commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,264 @@
+    # any combination of additional 'dot' digits pairs are accepted
+    # anything more than libFOO.so.digits.digits.digits
+    # should be seen as a member name outside normal expectations
+    exprs = [r'lib%s\.so\.[0-9]+[0-9.]*' % name,
```
Use raw f-strings here (and in the next line):
`fr'lib{name}\.so\.[0-9]+[0-9.]*'`

Mariatta commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,264 @@
+                continue
+            # "lib" is prefixed to emulate compiler name resolution,
+            # e.g., -lc to libc
+            base = 'lib%s.a' % name
```
Turn this to f-string:
`f'lib{name}.a'`

Mariatta commented on .gitignore:
```diff
@@ -1,5 +1,7 @@
 # Two-trick pony for OSX and other case insensitive file systems:
 # Ignore ./python binary on Unix but still look into ./Python/ directory.
```
Seems this comment should be together with line 5 (`/python`)

aixtools replied:
.gitignore changes made were only for my local environment. I guess I need to reset this and then add .gitignore to my local .gitignore so it gets removed from the PR

vstinner replied:
If you need to ignore more files on AIX, please propose a different PR for .gitignore ;-)

aixtools commented:
where do I add news and/or say "skip news".

Mariatta commented:
This will need a news entry, and you can add it using [blurb](https://pypi.org/project/blurb/)

Mariatta commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,264 @@
+                break
+        return info
+
+    p = Popen(["/usr/bin/dump", "-X%s" % aixABI(), "-H", file],
```
Use f-string:
`f"-X{aixABI()}"` instead of `"-X%s" % aixABI()`

aixtools replied:
If you insist. I have no clue what an f-string is, and will struggle - a lot - to read code using this syntax.

Sigh...

rest to come - next week then. You see it as simple, I see it as complex.

Mariatta commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,264 @@
+    and trailing "]"
+    """
+    # member names in the dumpH output are between square brackets
+    expr = r'\[(%s)\]' % expr
```
This can be made into raw f-strings:
`fr'[({expr})]'`

aixtools commented:
thx.

Too bad the file cannot just be added by using an editor. While packaging I do not have any python installed - so definitely in a chicken and egg circle here...

> You can install blurb from PyPI using pip. Alternatively, simply add blurb to a directory on your path. blurb’s only dependency is Python 3.5+.

aixtools commented:
p.s. how many days before CLA is reviewed again?

Mariatta commented:
I rechecked CLA :)

If you don't have blurb, the News entry can be added manually:
- Create a new file at  Misc/NEWS.d/next/ .... containing the news entry. You'll need to read blurb readme about the file name requirement.
- commit it

aixtools commented:
thx for the CLA update.

I had already found the directory and the readme. will work on that today.

Update: installed python3-3.5.2 on AIX, pip installed blurb - seemed to be working - but blurb insists on calling git - so that failed (my git depends on python2 and I have not tested having python2 and python3 installed in parallel. Been working on that, but not tested it yet.

Looked at: PRETTY_NAME="Debian GNU/Linux stretch/sid" (on power)

python3.5.1 - but no pip, not even via apt install pip
centos1611 - no python3 - and the latest python2 is 2.7.5 (really sad!)

So, when I am back home I'll start testing parallel python installs - as I do not see how I can fake a random number, or is it paired to a git commit number, etc..

Here is the text I would like to add - in case someone can tell me what the file name should be.

fyi: I also have an update to Modules/_ctypes/_ctypes.c (so that import RTLD_MEMBER from _ctypes would work) - and configure.ac so that a "HAVE_RTLD_MEMBER" gets defined (not perfect define name, but close enough I hope for discussion).

Shall I push those updates now - or wait until the blurb is also ready?

aixtools committed:
Add NEWS and last bits needed by configure and _ctypes

aixtools committed:
cannot start blurb with 'bpo-'

aixtools commented:
Little bit amazing to me that removing two words from the NEWS does not complete after two hours - while with the error (finding it took <3 minutes) the 3 other tests took about 30 minutes.

aixtools commented:
I just remembered - something I had originally worked in, but had to remove, was looking at LD_LIBRARY_PATH and LIBPATH. Need to add that back in. Update to come asap.

aixtools committed:
librpm.so is only 32-bit; compare cdll.LoadLibrary() and CDLL()

aixtools committed:
set LD_LIBRARY_PATH as preferred env variable

aixtools commented:
travis ended with this error: oci runtime error: exec failed: container_linux.go:265: starting container process caused "could not create session key: disk quota exceeded - so I hope someone will cause a repeat for the pass (expected).

I have tested localally in both 32-bit and 64-bit mode. The 64-bit mode returns. as expected:
michael@x071:[/data/prj/python/git/python-3.7.0.1]./python Lib/ctypes/util.py
None
libc.a(shr_64.o)
libbz2.a(libbz2.so.1)
<CDLL 'libc.a(shr_64.o)', handle d at 0x7000000001e7a20>
<CDLL 'libc.a(shr_64.o)', handle e at 0x7000000001e7a20>

aixtools commented:
Could someone please get the tests to re-run. The failure was established in one second - due to lack of disk space - nothing related to code.

![image](https://user-images.githubusercontent.com/10852597/33385988-44b0482a-d52a-11e7-8698-4e1267bbaa89.png)

Mariatta commented:
I'll close and re-open the PR, it'll trigger the CI.

aixtools commented:
thx. so, learning to wait for queues :)

(update: and now that the tests are passed - wait for someone to accept assignment and review!)

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    """
+    return sizeof(c_void_p) * 8
+
+def get_dumpH(file):
```
instead of "H", would you mind to replace it with the meaning of the option?

aixtools replied:
done

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+from ctypes._util import _last_version
+from subprocess import Popen, PIPE
+
+def aixABI():
```
PEP 8 suggests the name: aix_abi()

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+
+def aixABI():
+    """Internal support function:
+    return executable size - 32-bit, or 64-bit
```
"return executable size - 32-bit, or 64-bit"

I propose: "Get the executable bits: return 32 or 64"

aixtools replied:
adjusted

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+def aixABI():
+    """Internal support function:
+    return executable size - 32-bit, or 64-bit
+    This is vital to the search for suitable member in an archive
```
"This is vital to the search for suitable member in an archive"

I don't think that this sentence is useful. At soon as a function is used, it's useful ;-)

aixtools replied:
adjusted

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+        objects.append((object, get_info(p)))
+
+    p.stdout.close()
+    p.wait()
```
Please use "with p: ..." to automatically close stdout on error.

vstinner replied:
You may raise an exception if p.returncode is non-zero.

aixtools replied:
Again, this was based on recommendation from someone else. I'll reread the documentation. I hope
with Popen(....) as p:

is what you mean

If p.returncode is non-zero then there is 'nothing to find' and a return value of None is still ok.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    # 3. get info (lines starting with [0-9])
+
+    def get_object(p):
+        object = None
```
"object()" is a builtin function, you may rename the variable to "obj" to avoid name conflict.

aixtools replied:
done - i.e., renamed variables.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+        return info
+
+    p = Popen(["/usr/bin/dump", "-X%s" % aixABI(), "-H", file],
+        universal_newlines=True, stdout=PIPE)
```
You may redirect stderr to subprocess.DEVNULL, or to stdout (subprocess.STDOUT), to hide errors.

aixtools replied:
a) these options were chosen to be suitable for python2 as well as python3. Will have to read the documentation.
b) will redirect stderr to DEVNULL

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    return objects
+
+def get_shared(input):
+    """Internal support function: examine the get_dumpH() output and
```
I'm not sure that it's useful to mention "Internal" in each docstring, since the whole module is marked as private with the "_" prefix in its name.

aixtools replied:
That's what I need a mentor for. However, this stems from uncertainity about how this information is extracted and presented by programs than scan for """ ... """ blocks.

I'll delete the prefix.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    Note: the "[" and "]" characters that are part of dump -H output
+    are not removed here.
+    """
+    list = []
```
"list()" is a builtin type/function, use a different name like "shared" to avoid name conflict.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+from sys import executable
+from ctypes import c_void_p, sizeof
+from ctypes._util import _last_version
+from subprocess import Popen, PIPE
```
I would prefer to import only the module and use the full name. "from os import environ" coding style is rare in the stdlib. Replace replace "search" with "re.search".

aixtools replied:
OK. Long time ago - but I thought I was following some guidelines (maybe not PEP guidelines).

To be clear - replace ONLY search with re.search, or do you mean all related calls such as:
executable -> sys.executable, c_void_p and sizeof to ctypes.c_void_p and ctypes.sizeof

And regarding _last_version - ctypes._util._last_version (this routine was suggested by Martin, rather than a copy of the routine that it came from (and hence removed from util.py)

iirc from what I read before - the reason to not import the whole module had to do with memory consumption and/or performance (i.e., importing the whole module is (relatively) expensive.

So, awaiting your reply.

vstinner replied:
Please replace all imports.

> iirc from what I read before - the reason to not import the whole module had to do with memory consumption and/or performance (i.e., importing the whole module is (relatively) expensive.

I never heard anything like this.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    expr = r'\[(%s)\]' % expr
+    matches = list(filter(None, (search(expr, line) for line in lines)))
+    if len(matches) == 1:
+        return matches[0].group(1)
```
I suggest to write a regular loop to build matches, and store .group(1) in the matches, rather than the re.SRE_Match object.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+            list.append(line[line.index("["):-1])
+    return list
+
+def get_exactMatch(expr, lines):
```
PEP8: get_exact_match()

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+        member = get_exactMatch(expr, members)
+        if member:
+            return member
+    # 32-bit legacy names - both shr.o and shr4.o exist.
```
nitpick: i would put the comment block inside the else block, so indented.

aixtools replied:
looks more consistent

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+        # AIX 64-bit member is one of shr64.o, shr_64.o, or shr4_64.o
+        expr = r'shr4?_?64\.o'
+        member = get_exactMatch(expr, members)
+        if member:
```
"if member:" is redundant here, no? The function returns None if there is no match.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    else:
+        return get_legacy(members)
+
+def getExecLibPath_aix():
```
PEP8: get_exec_libpath_aix() (i'm not sure about the underscores :-) maybe get_exec_lib_path_aix, or get_execlib_path_aix!?)

aixtools replied:
shorter - get_libpaths()

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+    If no archive is found, look for a .so file.
+    """
+
+    def find_shared(paths, name):
```
Why not putting this function at the module level?

vstinner commented on Lib/ctypes/_util.py:
```diff
@@ -0,0 +1,15 @@
+        nums = []
+        try:
+            while parts:
+                nums.insert(0, int(parts.pop()))
```
I would prefer to iterate in the reverse order rather than modifying the list:

for part in reversed(parts):
   try: ...
   except ValueError: break

vstinner commented on Lib/ctypes/_util.py:
```diff
@@ -0,0 +1,15 @@
+                nums.insert(0, int(parts.pop()))
+        except ValueError:
+            pass
+        return nums or [maxsize]
```
I would prefer to return None if there no version, but replace None with something else *in the caller*.

aixtools replied:
The code for _last_version comes from util.py:

```
@@ -170,17 +180,6 @@ elif os.name == "posix":
     if sys.platform.startswith(("freebsd", "openbsd", "dragonfly")):

-        def _num_version(libname):
-            # "libxyz.so.MAJOR.MINOR" => [ MAJOR, MINOR ]
-            parts = libname.split(b".")
-            nums = []
-            try:
-                while parts:
-                    nums.insert(0, int(parts.pop()))
-            except ValueError:
-                pass
-            return nums or [sys.maxsize]
-
         def find_library(name):
             ename = re.escape(name)
             expr = r':-l%s\.\S+ => \S*/(lib%s\.\S+)' % (ename, ename)
@@ -199,7 +198,7 @@ elif os.name == "posix":
             res = re.findall(expr, data)
             if not res:
                 return _get_soname(_findLib_gcc(name))
-            res.sort(key=_num_version)
+            res.sort(key=_last_version)
             return os.fsdecode(res[-1])

```
I am reluctant to change it (already had one reviewer ask why I had even made _util.py).

And, as our way of finding, better, creating what we are looking through is different - I am more reluctant.

```
    exprs = [r'lib%s\.so\.[0-9]+[0-9.]*' % name,
        r'lib%s_?64\.so\.[0-9]+[0-9.]*' % name]
    for expr in exprs:
        versions = []
        for line in members:
            m = re.search(expr, line)
            if m:
                versions.append(m.group(0))
        if versions:
            return _last_version(versions, '.')
    return None
```

* The goal was to have a single _last_version() definition for both uses.
However, if this is confusing maybe there should be two versions (i.e., remove _util.py, restore num_version() to util.py and move (and adjust _last_version() within _aix.py

vstinner commented on Lib/ctypes/_util.py:
```diff
@@ -0,0 +1,15 @@
+from sys import maxsize
```
I don't think that a single function justify the creation of a whole module. Just put this code into the existing util.py

vstinner replied:
Why not putting this code in util.py directly? If it's really specific to AIX, maybe put it in _aix.py instead?

vstinner commented on Lib/ctypes/util.py:
```diff
@@ -80,6 +81,14 @@ def find_library(name):
+    import ctypes._aix as aix
+
+    def find_library(name):
+        return aix.find_library(name)
```
This can be written: find_library = aix.find_library, or simpler: from ctypes._aix import find_library

aixtools replied:
Yes indeed!

vstinner commented on Lib/ctypes/util.py:
```diff
@@ -324,6 +322,20 @@ def test():
+        # issue-26439 - fix broken test call for AIX
+        elif sys.platform.startswith("aix"):
+            from ctypes import CDLL
+            RTLD_MEMBER =  0x00040000
```
Please use os.RTDL_MEMBER instead of an hardcoded constant.

aixtools replied:
done.

vstinner commented on Misc/NEWS.d/next/Library/2017-11-24-08-35-43.bpo-26439.IC45_f.rst:
```diff
@@ -0,0 +1,10 @@
+Fix ctypes.util.find_library() for AIX
```
I'm not sure of this summary. In short, I understand that you *implemented* find_library() on AIX, not just "fixed".

aixtools replied:
Maybe it just needs to be implemented - I take that to be the "reviews" position.

However, ctypes.util.find_library() does exist for AIX - asis. And, this is broken.
The documentation makes clear that the behavior of find_library is platform-dependent.

So if you object to saying that this patch 'fixes' the published (aka existing feature of ctypes.util) because this patch 'implements' (as ctypes._aix.find_library(), just as "darwin" implements it as ctypes.macholib.dyld.dyld_find() (this was actually what I tried to emulate, just without a new sub-directory, and per much earlier reviews renamed it ctypes._aix to make clear it is not something to be imported automagically. Again - it is a platform dependency (and your suggestion to just import it into the namespace is excellent. Should have thought of that long ago.)

In closing - I see ctypes.util.find_library() as the API definition, and the current implementation is broken. This patch replaces that implementation with a new one - and this 'fixes' the API for AIX.
If the reviewers feel that the word 'fixed' is misplaced - I'll delete the line. Or, could/should I say I fixed the AIX implementation of find_library().

vstinner commented on Misc/NEWS.d/next/Library/2017-11-24-08-35-43.bpo-26439.IC45_f.rst:
```diff
@@ -0,0 +1,10 @@
+while legacy names e.g., find_library("c") returns libc.a(shr.o)
+or libc.a(shr_64.o) - depending on 32 or 64-bit operations.
+Include RTLD_MEMBER to mode to support AIX legacy library(member) names
+(Modules/_ctype/_ctype.c), ctypes/__init__.py and configure.ac)
```
Please credit yourself: add "Patch by ..." (your full name).

aixtools replied:
ok.

vstinner commented on Modules/_ctypes/_ctypes.c:
```diff
@@ -5582,6 +5582,12 @@ PyInit__ctypes(void)
+#endif
+#if HAVE_DECL_RTLD_NOW
+    PyModule_AddObject(m, "RTLD_NOW", PyLong_FromLong(RTLD_NOW));
+#endif
```
Please move this to Modules/posixmodule.c. We already have many os.RTLD_* constants there... like os.RTLD_NOW ;-)

aixtools replied:
ok. as above. In my previous convos about this - they were being put into _ctypes - so I just used that convention.

vstinner commented on Lib/ctypes/__init__.py:
```diff
@@ -338,6 +338,16 @@ def __init__(self, name, mode=DEFAULT_MODE, handle=None,
             flags |= _FUNCFLAG_USE_LASTERROR
+        if _sys.platform.startswith("aix"):
+            """When the name contains ".a(" and ends with ")",
+               asin "libFOO.a(libFOO.so)" this is taken to be an
```
asin: typo for as in?

aixtools replied:
Yes, I frequently type "asin" as a single world in presentations I give. I'll change it to two words here.

vstinner reviewed (commented):
Enjoy a first round of review ;-)

The overall change LGTM, but I have sent a long list of requests ;-)

Mariatta reviewed (commented):
My comments are only about using f-strings in _aix.py :) I believe this patch will only go to Python >= 3.7. 
Also, should add yourself to Misc/ACKS.

aixtools commented:
@victor Thx for the long list. 

Also @Mariatta for the part I hate most - reg_expr_parsing.

p.s. - As I had no clue what replying to this as an email would do - I am editing this reply to something shorter.

re: no backport - I'll try to respond below to victor - and I hope without spilling any more milk.

vstinner commented:
> Only in >= 3.7...

Right. You are writing code for the master branch, so you can new all new shiny Python features :-) Cool, isn't it?

This change is not going to be backported. Implement find_library() on AIX is like a new feature, and we don't accept new features in stable branches. If we want to do that, usually, what people do is to then maintain a third party backport on PyPI. There are many of such backports as subprocess, faulthandler, tracemalloc, etc.

aixtools commented:
> This change is not going to be backported. Implement find_library() on AIX is like a new feature, and we don't accept new features in stable branches. 
Ah - my confusion. As I can call find_library() from AIX - I thought the feature was there, but the implementation of that feature was broken for AIX.
> If we want to do that, usually, what people do is to then maintain a third party backport on PyPI. There are many of such backports as subprocess, faulthandler, tracemalloc, etc.
I am mainly confused about the 'we' in above - who is 'we'?

And, while I feel it is more elegant to have the AIX code, like most of the MACOS code, to be in a separate file, rather than forced into the file util.py - is it a 'feature' because it is a separate file?

To me: feature is a function performed, and an implementation is merely the how. What got me started on this was some python code that failed because they were making calls such as:

libc = ctypes.cdll.LoadLibrary(ctypes.util.find_library("c")) - and - LIBPAM = CDLL(find_library('pam'))

And my goal was to be sure that these would be portable to AIX. Easy for the programmer.

As is both libc and LIBPAM are assigned whatever ctypes.cdll.LoadLibrary(None) and CDLL(None) return.

In closing - if this, from your eyes, is a 'feature' change, I accept there is nothing I can do or say about it. But I do hope you can at least acknowledge that, from my eyes, this is a correction of a long ignored bug.

Moving forward: many thanks for the clarification; later I will want to discuss how to engage with PyPI to have it somewhere. (Now I need some coffee to drown my sorrows :smile:)

aixtools committed:
ignore further changes to .gitignore (local changes for build)
 add generated configure and pyconfig.h.in after change to configure.ac

aixtools committed:
Add some AIX specific tests

aixtools committed:
correct a typo

aixtools committed:
changes made based on review comments

aixtools committed:
small corrections

vadmium commented on Lib/ctypes/__init__.py:
```diff
@@ -338,6 +338,16 @@ def __init__(self, name, mode=DEFAULT_MODE, handle=None,
+            from _ctypes import RTLD_MEMBER,RTLD_NOW
+            if name and name.endswith(")") and ".a(" in name:
+                # RTLD_MEMBER = 0x00040000
+                mode |= ( RTLD_MEMBER | RTLD_NOW )
```
Why did you add RTLD_NOW back again? Wasn’t my explanation in the documentation enough?

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+def get_shared(input):
+    """Internal support function: examine the get_dumpH() output and
+    return a list of all shareable objects indicated in the output the
+    character "[" is used to strip off the path information.
```
Previous comment still stands: <https://bugs.python.org/issue26439#msg278675>

vadmium commented on Lib/ctypes/util.py:
```diff
@@ -80,6 +81,14 @@ def find_library(name):
 
+if sys.platform.startswith("aix"):
+    # find .so members in .a files
+    # using dump loader header information + sys.
```
What is “sys”?

aixtools replied:
The same thing it is - where I copied an example from.

michael@x071:[/data/prj/python/git/python-3.7.0.1]find . -name \*.py | xargs grep "sys.platform"
./Lib/_bootlocale.py:if sys.platform.startswith("win"):
./Lib/_bootlocale.py:            if not result and sys.platform == 'darwin':
./Lib/_osx_support.py:    if (sys.platform == 'win32') and (ext != '.exe'):
./Lib/_pyio.py:if sys.platform in {'win32', 'cygwin'}:
./Lib/asyncio/__init__.py:if sys.platform == 'win32':
./Lib/asyncio/__init__.py:if sys.platform == 'win32':  # pragma: no cover
./Lib/asyncio/base_events.py:                reuse_address = os.name == 'posix' and sys.platform != 'cygwin'
./Lib/asyncio/base_events.py:                reuse_address = os.name == 'posix' and sys.platform != 'cygwin'
./Lib/asyncio/test_utils.py:if sys.platform == 'win32':  # pragma: no cover
./Lib/asyncio/unix_events.py:if sys.platform == 'win32':  # pragma: no cover
./Lib/asyncio/unix_events.py:        if is_socket or (is_fifo and not sys.platform.startswith("aix")):
./Lib/asyncio/windows_utils.py:if sys.platform != 'win32':  # pragma: no cover
./Lib/code.py:                       (sys.version, sys.platform, cprt,
./Lib/ctypes/__init__.py:if _os.name == "posix" and _sys.platform == "darwin":
./Lib/ctypes/__init__.py:        if _sys.platform.startswith("aix"):
./Lib/ctypes/__init__.py:elif _sys.platform == "cygwin":
./Lib/ctypes/util.py:if os.name == "posix" and sys.platform == "darwin":
./Lib/ctypes/util.py:if sys.platform.startswith("aix"):
./Lib/ctypes/util.py:    if sys.platform == "sunos5":
./Lib/ctypes/util.py:    if sys.platform.startswith(("freebsd", "openbsd", "dragonfly")):
./Lib/ctypes/util.py:    elif sys.platform == "sunos5":
./Lib/ctypes/util.py:        if sys.platform == "darwin":
./Lib/ctypes/util.py:        elif sys.platform.startswith("aix"):
...

"sys.platform" seems to be fairly frequent

michael@x071:[/data/prj/python/git/python-3.7.0.1]find . -name \*.py | xargs grep "sys.platform" | wc -l
569

So, please explain your question. I thought 'sys' was well understood. Have I misused it in some way?

vadmium replied:
In the comment “using dump loader header information + sys”, I did not realize that you are referring to the _sys_ Python module, nor the _sys.platform_ string. I don’t understand how you can use _sys.platform_ “to find members in files”. You are only using it to determine if the code is called on AIX.

vadmium commented on Lib/ctypes/util.py:
```diff
@@ -324,6 +322,20 @@ def test():
+            from ctypes import CDLL
+            RTLD_MEMBER =  0x00040000
+            # print("crypto\t:: %s" % cdll.LoadLibrary(find_library("crypto")))
+            if (sys.maxsize < 2**32):
```
Round brackets have re-appeared

aixtools replied:
Do you mean this should be:
if {sys.maxsize < 2**32}:

Confused! by comment.

vadmium replied:
It would be easier to read (IMO) without any kind of brackets. I removed them when I modified your patch (https://bugs.python.org/review/26439/diff2/18236:18714/Lib/ctypes/util.py), and they stayed removed in your updated aix-library.161004.patch. See also test_loading.py.

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+        if member:
+            return member
+    # 32-bit legacy names - both shr.o and shr4.o exist.
+    # shr.o is the preffered name so we look for shr.o first
```
https://bugs.python.org/issue26439#msg278675

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,265 @@
+
+    Before the GNU convention became the standard scheme regardless of
+    binary size AIX packagers used GNU convention "as-is" for 32-bit
+    archive members but used an "distinguishing" name for 64-bit members.
```
https://bugs.python.org/issue26439#msg278675

vadmium reviewed (commented):
@vstinner, the biggest thing that disqualified this as a bugfix for me was the change in documented behaviour of CDLL, now parsing the archive(member) format and adding RTLD_MEMBER. We discussed this a lot in Michael’s bug threads, but I don’t think he was happy with my view.

aixtools committed:
work on this as a seperate issue - def test_find()

aixtools committed:
Rather than import three functions from re, import all of re
 and call the routines by name, e.g., re.search, re.match, re.escape

vadmium commented on Lib/ctypes/test/test_loading.py:
```diff
@@ -13,6 +13,11 @@ def setUpModule():
+        if (sys.maxsize < 2**32):
+            libc_name = "libc.a(shr.o)"
+        else:
+            libc_name = "libc.a(shr_64.o)"
```
Is this really needed? Won’t your find_library("c") implementation find these?

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,312 @@
+        universal_newlines=True, stdout=PIPE, stderr=DEVNULL) as p:
+            ld_header = get_ld_header(p)
+            if ld_header:
+                ldr_headers.append((ld_header, get_ld_header_info(p)))
```
Since you now only return one item, would it be better to just return the tuple or None, without building a list?

``` python
if ld_header:
    return (ld_header, get_ld_header_info(p))
else:
    return None
```

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,312 @@
+
+def get_shared(ld_headers):
+    """
+    extract a the shareable objects from ld_headers
```
Extract the . . . [drop _a_]

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,312 @@
+    return libpaths
+
+def find_library(name):
+    """AIX implemantation of ctypes.util.find_library()
```
Spelling: _implementation_

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -1,19 +1,54 @@
-rather than as separate, detailed if: sections in utils.py
+"""
+Lib/ctypes.util.find_library() support for AIX
+Similar approach as done for Darwin support by using seperate files
```
separate

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -1,19 +1,54 @@
+2.9 SVR4 linking affinity
+Nowadays, there are two major object file formats used by the operating systems:
+XCOFF: The COFF enhanced by IBM and others. The original COFF (Common
+Object FIle Format) was the base of SVR3 and BSD 4.2 systems.
```
Lowercase _i_ in _File_?

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -1,19 +1,54 @@
+library object is referred to as the "member".
+
+For dlopen() on AIX (read initAndLoad()) the calls are similiar.
+Default activity is achived by not providing any path information. When path
```
Spelling: _achieved_ (or _archived_?)

vadmium commented on Lib/ctypes/_aix.py:
```diff
@@ -1,19 +1,54 @@
+
+For dlopen() on AIX (read initAndLoad()) the calls are similiar.
+Default activity is achived by not providing any path information. When path
+information is provided dlopen() does not search any alturnate directories.
```
Spelling: _alternate_ (or to avoid the Americanism, _alternative_ or _other_)

aixtools committed:
remove changes to test_loading.py - for another day

aixtools committed:
Responding to reviewer comments
 Restored a "while True: loop to be sure file was being read to the end
  otherwise, only the first entry was being returned.
 Corrected spelling errors
 Clarified some comments
 Modified some (temporary) tests in util.py

aixtools committed:
made changes per reviewer comments/suggestions

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,323 @@
+naming style.
+"""
+__author__ = "Michael Felt <aixtools@felt.demon.nl>"
+__version__ = "1.0.0"
```
I dislike having versions in stdlib modules, especially in submodules. IMHO the Python version is enough, since the only ship the stdlib with Python. I suggest to remove __version.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,323 @@
+    # 2. If "INDEX" in occurs in a following line - return ld_header
+    # 3. get info (lines starting with [0-9])
+
+    def get_ld_header(p):
```
Nested functions are slow, since they are created each time you call get_ld_headers(). Would it be possible to put them at the module scope?

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,323 @@
+
+    ldr_headers = []
+    p = Popen(["/usr/bin/dump", "-X%s" % aix_abi(), "-H", file],
+        universal_newlines=True, stdout=PIPE, stderr=DEVNULL)
```
Would you mind to use "with p:" here to close the subprocess in case of error?

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,323 @@
+        universal_newlines=True, stdout=PIPE, stderr=DEVNULL)
+    # be sure to read to the end-of-file - getting all entries
+    while True:
+            ld_header = get_ld_header(p)
```
Indentation should be fixed.

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,323 @@
+    Return executable bit size - 32 or 64
+    Used to filter the search in an archive by size, e.g., -X64
+    """
+    return sizeof(c_void_p) * 8
```
Why not putting directly the result into a constant? AIX_ABI = sizeof(c_void_p) * 8. sizeof() is fast and cheap, and the result is not going to change :-)

vstinner commented on Lib/ctypes/_aix.py:
```diff
@@ -0,0 +1,323 @@
+    If no archive+member pair is found, look for a .so file.
+    """
+
+    def find_shared(paths, name):
```
Again, would it be possible to put this function at the module scope for performance?

vstinner reviewed (commented):
f-strings are not mandatory. I propose to go without f-string, and if @Mariatta really want to get f-string in ctypes, maybe @Mariatta can write a more global f-string PR for the whole ctypes module, once the AIX code landed?

@aixtools: I rewiewed the latest version of your PR, and it's almost good to be merged. IMHO the remaining blocker issue the the creation of the _util.py file (I would prefer to not create such file for a single function, I proposed solutions).

@aixtools: It would be sad to lose this work just because of coding style. As I wrote, I'm ok to take your PR without f-string. It was more a proposal than a requirement.

Thank your for your work, and come on, it's almost done! It would be great to finally fix this very old AIX issue in Python 3.7!

aixtools committed:
updates per reviewer requests
